logstreams/src/main/java/io/camunda/zeebe/logstreams/log/LogStreamWriter.java

```diff
@@ -22,14 +39,65 @@ public interface LogStreamWriter {
   default long tryWrite(final LogAppendEntry appendEntry) {
     return tryWrite(appendEntry, LogEntryDescriptor.KEY_NULL_VALUE);
   }
+  /** {@inheritDoc} */
+  default long tryWrite(final LogAppendEntry appendEntry, final long sourcePosition) {
+    return tryWrite(Collections.singletonList(appendEntry), sourcePosition);
+  }
 
   /**
-   * Attempts to write the event to the underlying stream.
+   * Attempts to write the events to the underlying stream. This method is atomic, either all events
+   * are written, or none are.
    *
-   * @param appendEntry the entry to write
-   * @param sourcePosition a back-pointer to the record whose processing created this entry
-   * @return the event position, a negative value if fails to write the event, or 0 if the value is
-   *     empty
+   * @param appendEntries a set of entries to append; these will be appended in the order
+   * @return the last (i.e. highest) event position, a negative value if fails to write the events,
+   *     or 0 if the batch is empty
+   */
+  default long tryWrite(final LogAppendEntry... appendEntries) {
+    return tryWrite(Arrays.asList(appendEntries));
+  }
+
+  /**
+   * Attempts to write the events to the underlying stream. This method is atomic, either all events
+   * are written, or none are.
+   *
+   * @param appendEntries a set of entries to append; these will be appended in the order in which
+   *     the collection is iterated.
+   * @return the last (i.e. highest) event position, a negative value if fails to write the events,
+   *     or 0 if the batch is empty
+   */
+  default long tryWrite(final Iterable<? extends LogAppendEntry> appendEntries) {
+    return tryWrite(appendEntries, LogEntryDescriptor.KEY_NULL_VALUE);
+  }
+
+  /**
+   * Attempts to write the events to the underlying stream. This method is atomic, either all events
+   * are written, or none are.
+   *
+   * @param appendEntries a set of entries to append; these will be appended in the order in which
+   *     the collection is iterated.
+   * @param sourcePosition a back-pointer to the record whose processing created these entries
+   * @return the last (i.e. highest) event position, a negative value if fails to write the events,
+   *     or 0 if the batch is empty
+   */
+  default long tryWrite(
+      final Iterable<? extends LogAppendEntry> appendEntries, final long sourcePosition) {
+    return tryWrite(
```

Comment:
:wrench: Looking at this, where we create a stream then convert it to a list...this method is mostly there for the engine, right? The `ProcessResultBuilder` or `RecordBatch`? The `RecordBatch` wraps an underlying list, can't we just...bypass this? Pass the list directly so we don't support `Iterable<?>`? In the end, we want anyway the items as a collection, so this is more annoying than anything, no? If we're worried about immutability and passing the original list reference, we could just wrap it in `Collections.unmodifiableList` (this would also allow us to remove the pesky cast to `(List<LogAppendEntry>)` in the Sequencer).

Marking this a suggestion because I'm not 100% sure of the usage, and it's pretty minor.

Reply:
> this method is mostly there for the engine, right

Yes, that's pretty much the only one passing and iterable. Let me check again if I can modify the interface there, it'd be great to not need iterables :+1: